README.md
Introduction
============

This repository is a fork of PEAR::Numbers_Words as composer package with PSR-0 compliance.

Credits to all previous authors.

```
With PEAR::Numbers_Words class you can change an integer number
to simple words. This can be usefull when you need to spell a currency
value e.g. on an invoice.

You can choose between several languages (language files are located
in src/Numbers/Words/Locale directory).

BTW: if you need to change number to currency, use money_format()
PHP function (available since 4.3 (not yet released)). But you must
know that locale LC_MONETARY rules are sometimes unclear.
```

Getting started
---------------

First you need to install lla/numbers_words PEAR package.
You can do it (as root) with:

```php
composer install ncrypthic/numbers
```

In your php script you need to ```use Numbers\Words```header file:

```php
use \LLA\Numbers\Words;
```

Then you can call ```Numbers_Words::fromWords()``` function with two
arguments: integer number (can be a string with digits) and
optional locale name (default is *en_US*):

```php
use \LLA\Numbers\Words;

$num   = 12340000000;
$words = Words::fromWords($num,"en_GB");
echo "Num $num in British English is '<b>$words</b>'<p>\n";
```

For  this would display:
```
Num 12340000000 in British English is '<b>twelve thousand million three hundred forty million</b>'<p>
```

There are avaibale the following modules (called by locale name,
in alphabetical order):

  az     - Azerbaijani language.
           Author: Shahriyar Imanov

  bg     - Bulgarian language (in WIN-1251 charset).
           Author: Kouber Saparev

  cs     - Czech language.
           Author: Petr 'PePa' Pavel

  de     - German language.
           Author: Piotr Klaban

  dk     - Danish language.
           Author: Jesper Veggerby

  en_100 - Donald Knuth number naming system, in English language.
           Author: Piotr Klaban

  en_GB  - British English notation of numbers, where
           one billion is 1000000 times one million.
           1000 times million is just 'thousand million' here.
           I do not use a word billiard here, because
           English people do not use it often, and even could not know it.
           Author: Piotr Klaban

  en_US  - American English notation of numbers, where
           one billion is 1000 times one million
           Author: Piotr Klaban

  es     - Spanish (Castellano) language.
           Author: Xavier Noguer

  es_AR  - Argentinian Spanish language.
           Author: Martin Marrese

  et     - Estonian language.
           Author: Erkki Saarniit

  fr     - French language.
           Author: Kouber Saparev

  fr_BE  - French (Belgium) language.
           Author: Kouber Saparev, Philippe Bajoit

  he     - Hebrew language.
           Author: Hadar Porat

  hu_HU  - Hungarian language.
           Author: Nils Homp

  id     - Indonesia language.
           Authors: Ernas M. Jamil, Arif Rifai Dwiyanto

  it_IT  - Italian language.
           Authors: Filippo Beltramini, Davide Caironi

  lt     - Lithuanian language.
           Author: Laurynas Butkus

  nl     - Dutch language.
           Author: WHAM van Dinter

  pl     - Polish language (in an internet standard charset ISO-8859-2)
           Author: Piotr Klaban

  pt_BR  - Brazilian Portuguese language.
           Authors: Marcelo Subtil Marcal, Mario H.C.T., Igor Feghali

  ro_RO  - Romanian language.
           Author: Bogdan Stancescu	

  ru     - Russian language.
           Author: Andrey Demenev

  sv     - Swedish language.
           Author: Robin Ericsson

  tr_TR  - Turkish language.
           Author: Shahriyar Imanov

** What if numbers have fraction part?

You can split the number by the coma or dot. The example
function was provided by Ernas M. Jamil (see below).
I do not know if the splitting and concatenating numbers
should be supported by Numbers_Words ... Does each language
spell numbers with a 'coma'/'koma'? What do you think?

```php
use \LLA\Numbers\Words;

function num2word($num, $fract = 0) {

    $num = sprintf("%.".$fract."f", $num);
    $fnum = explode('.', $num);

    $ret =  Words::fromNumber($fnum[0],"id");
    if(!$fract) return $ret;

    $ret .=  ' koma '; // point in english
    $ret .= Words::fromNumber($fnum[1],"id");

    return $ret;
}
```

** How to convert decimal part and not fraction part of the currency value?

Rob King send me a patch that would allow to leave fraction part in digits.
I.e. you can convert 31.01 into 'thirty-one pounds 01 pence':

```php
use \LLA\Numbers\Words;
use \LLA\Numbers\Words\Locale\en\GB;

$obj = new GB();
$convert_fraction = false;
print $obj->toCurrencyWords('GBP', '31', '01', $convert_fraction) . "\n";

```

** How to write new Language Files:

Just copy existing en_US or en_GB etc. file into src/LLA/Numbers/Words/Locale/{your_country/locale code}.php
and translate digits, numbers, tousands to your language. Then please send it
to the author to the address makler@man.torun.pl.

** Credits

All changes from other people are desrcribed with details in ChangeLog.
There are also names of the people who send me patches etc.
Authors of the language files are mentioned in the language files directly
as the author.


ChangeLog
2015-09-01
  * Refactor to PSR-0 compliance
  * Create LocaleInterface for Locale classes
  * Rename all methods with prefix '_' to camelCase with private method
  * Replace all `var` keyword with property access type
  (private|public|protected)
  * Rename```Words:_toWords()``` to ``Words::fromNumber()```
  * Update tests class with new namespace

2014-11-03
  * Fixed bug #20435: Missing files in PEAR archive [kouber]
  * Fixed French plural handling for "million" and suffixes above [Olivier Brunel]

2014-10-24
  * Fixed bug #19453: Incorrect spelling of Hungarian numbers [kouber]
  * Fixed bug #19543: Better handling of decimal mark and thousands separators [kouber]
  * Fixed bug #19855: Do not emit E_NOTICE when calling toWords() statically [cweiske]
  * Fixed pl unicode [Jakub Roszkiewicz]
  * Fixed ru_RU currency codes [Vital Leshchyk]
  * Fixed tr_TR unicode [Shahriyar Imanov]
  * Added en_IN - Indian English [Abhinav Nayak]
  * Added lv - Latvian [Andrius]
  * Added ro_RO - Romanian [Bogdan Stancescu]
  * Added ua - Ukrainian [Vital Leshchyk]
  * Unify locale loading code [cweiske]
  * Use PEAR class-to-filename convention [cweiske]
  * Changed class names and locations of locale files

2012-07-01
  * Drop PHP 4
  * Drop PEAR dependency
  * Misc fixes and cleanup in Turkish driver

2012-02-22
	* Numbers_Words: Version 0.16.4 released.
	* Re-licensed to PHP 3.01 (ifeghali)

2011-12-10
	* Bug #12512 Windows incorrect declaration of toWords() (ifeghali)
	* Bug #17373 bug in toWords function for hungarian translation (kouber)
	* Bug #17392 bugs in toWords function for hungarian translation (ifeghali)
	* Bug #17754 Patch: Using explode instead of split (ifeghali)
	* Request #17763 Independent tr_TR Lang support (ifeghali)
	* Bug #17794 Patch: avoiding split in es_MX file (ifeghali)
	* Bug #18501 incorrect declaration of toWords() in lang.fr.php (kouber)
	* Bug #19013 remove error_reporting (for PEAR QA team) (ifeghali)

2010-05-10
	* Numbers_Words: Version 0.16.2 released.
	* Words.php: incorrect rounding of cents (Bug #16733).
	* Words/lang.pl.php: improper use of substr() in Polish extension (Bug #16435).
	* Words/lang.hu_HU.php: wrong spelling of 40 in Hungarian extension (Bug #16667) (Vlajos).
	* tests/AllTests.php: switch to TextUI/TestRunner.php framework (Daniel O'Connor).

2009-03-14
	* Numbers_Words: Version 0.16.1 released.
	* Words.php: better cast/sanitize to int.
	* Words/lang.it_IT.php: fixed incorrect trimming of vowels (Bug #16019) (Filippo Beltramini).
	* tests/ItalianTest.php: incorrect trimming, reveling Bug #16019 (Lorenzo Alberton).
	* tests/*: fix line endings.

2008-03-10
	* Numbers_Words: Version 0.16.0 released.
	* Words.php: fixed wrong getLocales() Docblock  bug (#10744) (Daniel O'Connor).
	* Words.php: fixed bogus non-int parsing (#14988) (Daniel O'Connor).
	* Words/lang.cs.php: fixed bad trascription for 101 numbers and 102 thousands bug (#14667) (Daniel O'Connor).
	* Words/lang.fr_BE.php: fixed a bug involving incorrect values for 80 and 81 (Kouber Saparev).
	* Words/lang.pt_BR.php: bug fix (#15260) (Igor Feghali).
	* tests/*: Swap to PHPUnit 3 (Daniel O'Connor).
	* Numerous language unit tests added (bg, en_100, en_gb, en_us, fr, fr_BE, de, it, pt_BR).
	* Fix numerous coding standard issues (bug #15145).

2006-06-12
	* Numbers_Words: Version 0.15.0 released.
	* Words/lang.nl.php: New language file: nl for Dutch (Thanks to WHAM van Dinter)
	* Words/lang.et.php: changed locale name for Estonian language from 'ee' to 'et' (#5618).
	* Words/lang.ru.php: fixed bug in Russian translation (if x modulo 10 was equal to 0) (#6654).
	* Words.php: new Numbers_Words::getLocales() method (Thanks to Bertrand Gugger).
	* tests/test-numbers-words.php: updated to use getLocales().
	* package.xml: updated to version 2.0.

2005-09-19
	* Numbers_Words: Version 0.14.0 released.
	* Words/lang.*.php: added $convert_fraction optional argument that allow to not convert fraction part in toCurrencyWorkds() method.
	* Words/lang.en_GB.php: fixed plural of pence; toCurrency() method added.

2005-03-09
	* Numbers_Words: Version 0.13.1 released.
	* Words.php: Fixed a problem with no leading zero in fraction of the currency value (reported by geekdom at gmail dot com).

2005-02-28
	* Numbers_Words: Version 0.13.0 released.
	* support for a new currency ILS (or NIS) for New Israeli Sheqel

2005-02-15
	* Words/lang.cs.php: New language file: cs for Czech (Thanks to Petr 'PePa' Pavel)

2005-02-03
	* Words/lang.he.php: New language file: he for Hebrew (Thanks to Hadar Porat)

2005-01-11
	* Numbers_Words: Version 0.12.0 released.
	* Words/lang.fr_BE.php: New language file: fr_BE for French (Belgium) (Thanks to Philippe Bajoit).

2004-10-25
	* Fixed spell mistake in some files - Britich to British (Thanks to Christian Weiske)

2004-10-22
	* Numbers_Words: Version 0.11.0 released.
	* Words/lang.pt_BR.php: toCurrencyWords() support in pt_BR translation (Thanks to Mario H.C.T).

2004-09-30
	* Numbers_Words: Version 0.10.1 released.
	* Words/lang.de.php: Small bug fix to German translation (the one thousand) (Thanks to neuner at futureweb dot at for the fix).
	* tests/test-numbers-words.php: Changed test number.

2004-09-20
	* Numbers_Words: Version 0.10.0 released.
	* Words/lang.dk.php: New language file: dk for Danish (Author: Jesper Veggerby)

2004-08-26
	* Numbers_Words: Version 0.9.0 released.
	* Words/lang.hu_HU.php: New language file: hu_HU for Hungarian (Author: Nils Homp)
	* Words/lang.id.php: Updating author list of the Indonesian translation (Arif Rifai Dwiyanto translated
	into Indonesian first but I missed his file by mistake).

2004-08-09
	* Numbers_Words: Version 0.8.1 released.
	* Words/lang.ru.php: fixed two bugs (missing t in 400, and wrong function call) in Russian translation (thanks
	to Anton Funk and igorag for reporting, and Andrey Demenev for fixing)

2004-06-17
	* Numbers_Words: Version 0.8 released.
	* Words/lang.lt.php: New file for Lithuanian language.

src/Numbers/Words/Locale/fr.php
<?php
/**
 * Numbers_Words
 *
 * PHP version 5
 *
 * Copyright (c) 1997-2006 The PHP Group
 *
 * This source file is subject to version 3.01 of the PHP license,
 * that is bundled with this package in the file LICENSE, and is
 * available at through the world-wide-web at
 * http://www.php.net/license/3_01.txt
 * If you did not receive a copy of the PHP license and are unable to
 * obtain it through the world-wide-web, please send a note to
 * license@php.net so we can mail you a copy immediately.
 *
 * @category Numbers
 * @package  Numbers_Words
 * @author   Kouber Saparev <kouber@php.net>
 * @license  PHP 3.01 http://www.php.net/license/3_01.txt
 * @version  SVN: $Id$
 * @link     http://pear.php.net/package/Numbers_Words
 */

/**
 * Include needed files
 */
namespace Numbers\Words\Locale;

use Numbers\Words;
use Numbers\Words\LocaleInterface;

/**
 * Class for translating numbers into French.
 *
 * @category Numbers
 * @package  Numbers_Words
 * @author   Kouber Saparev <kouber@php.net>
 * @license  PHP 3.01 http://www.php.net/license/3_01.txt
 * @link     http://pear.php.net/package/Numbers_Words
 */
class fr extends Words implements LocaleInterface
{

    /**
     * Locale name.
     * @var string
     * @access public
     */
    var $locale = 'fr';

    /**
     * Language name in English.
     * @var string
     * @access public
     */
    var $lang = 'French';

    /**
     * Native language name.
     * @var string
     * @access public
     */
    var $lang_native = 'Français';

    /**
     * The words for some numbers.
     * @var string
     * @access private
     */
    var $_misc_numbers = array(
        10=>'dix',      // 10
            'onze',     // 11
            'douze',    // 12
            'treize',   // 13
            'quatorze', // 14
            'quinze',   // 15
            'seize',    // 16
        20=>'vingt',    // 20
        30=>'trente',   // 30
        40=>'quarante', // 40
        50=>'cinquante',// 50
        60=>'soixante', // 60
       100=>'cent'      // 100
    );


    /**
     * The words for digits (except zero).
     * @var string
     * @access private
     */
    var $_digits = array(1=>"un", "deux", "trois", "quatre", "cinq", "six", "sept", "huit", "neuf");

    /**
     * The word for zero.
     * @var string
     * @access private
     */
    var $_zero = 'zéro';

    /**
     * The word for infinity.
     * @var string
     * @access private
     */
    var $_infinity = 'infini';

    /**
     * The word for the "and" language construct.
     * @var string
     * @access private
     */
    var $_and = 'et';

    /**
     * The word separator.
     * @var string
     * @access private
     */
    var $_sep = ' ';

    /**
     * The dash liaison.
     * @var string
     * @access private
     */
    var $_dash = '-';

    /**
     * The word for the minus sign.
     * @var string
     * @access private
     */
    var $_minus = 'moins'; // minus sign

    /**
     * The plural suffix (except for hundred).
     * @var string
     * @access private
     */
    var $_plural = 's'; // plural suffix

    /**
     * The suffixes for exponents (singular).
     * @var array
     * @access private
     */
    var $_exponent = array(
        0 => '',
        3 => 'mille',
        6 => 'million',
        9 => 'milliard',
       12 => 'trillion',
       15 => 'quadrillion',
       18 => 'quintillion',
       21 => 'sextillion',
       24 => 'septillion',
       27 => 'octillion',
       30 => 'nonillion',
       33 => 'decillion',
       36 => 'undecillion',
       39 => 'duodecillion',
       42 => 'tredecillion',
       45 => 'quattuordecillion',
       48 => 'quindecillion',
       51 => 'sexdecillion',
       54 => 'septendecillion',
       57 => 'octodecillion',
       60 => 'novemdecillion',
       63 => 'vigintillion',
       66 => 'unvigintillion',
       69 => 'duovigintillion',
       72 => 'trevigintillion',
       75 => 'quattuorvigintillion',
       78 => 'quinvigintillion',
       81 => 'sexvigintillion',
       84 => 'septenvigintillion',
       87 => 'octovigintillion',
       90 => 'novemvigintillion',
       93 => 'trigintillion',
       96 => 'untrigintillion',
       99 => 'duotrigintillion',
      102 => 'trestrigintillion',
      105 => 'quattuortrigintillion',
      108 => 'quintrigintillion',
      111 => 'sextrigintillion',
      114 => 'septentrigintillion',
      117 => 'octotrigintillion',
      120 => 'novemtrigintillion',
      123 => 'quadragintillion',
      126 => 'unquadragintillion',
      129 => 'duoquadragintillion',
      132 => 'trequadragintillion',
      135 => 'quattuorquadragintillion',
      138 => 'quinquadragintillion',
      141 => 'sexquadragintillion',
      144 => 'septenquadragintillion',
      147 => 'octoquadragintillion',
      150 => 'novemquadragintillion',
      153 => 'quinquagintillion',
      156 => 'unquinquagintillion',
      159 => 'duoquinquagintillion',
      162 => 'trequinquagintillion',
      165 => 'quattuorquinquagintillion',
      168 => 'quinquinquagintillion',
      171 => 'sexquinquagintillion',
      174 => 'septenquinquagintillion',
      177 => 'octoquinquagintillion',
      180 => 'novemquinquagintillion',
      183 => 'sexagintillion',
      186 => 'unsexagintillion',
      189 => 'duosexagintillion',
      192 => 'tresexagintillion',
      195 => 'quattuorsexagintillion',
      198 => 'quinsexagintillion',
      201 => 'sexsexagintillion',
      204 => 'septensexagintillion',
      207 => 'octosexagintillion',
      210 => 'novemsexagintillion',
      213 => 'septuagintillion',
      216 => 'unseptuagintillion',
      219 => 'duoseptuagintillion',
      222 => 'treseptuagintillion',
      225 => 'quattuorseptuagintillion',
      228 => 'quinseptuagintillion',
      231 => 'sexseptuagintillion',
      234 => 'septenseptuagintillion',
      237 => 'octoseptuagintillion',
      240 => 'novemseptuagintillion',
      243 => 'octogintillion',
      246 => 'unoctogintillion',
      249 => 'duooctogintillion',
      252 => 'treoctogintillion',
      255 => 'quattuoroctogintillion',
      258 => 'quinoctogintillion',
      261 => 'sexoctogintillion',
      264 => 'septoctogintillion',
      267 => 'octooctogintillion',
      270 => 'novemoctogintillion',
      273 => 'nonagintillion',
      276 => 'unnonagintillion',
      279 => 'duononagintillion',
      282 => 'trenonagintillion',
      285 => 'quattuornonagintillion',
      288 => 'quinnonagintillion',
      291 => 'sexnonagintillion',
      294 => 'septennonagintillion',
      297 => 'octononagintillion',
      300 => 'novemnonagintillion',
      303 => 'centillion'
        );

    /**
     * Split a number to groups of three-digit numbers.
     *
     * @param mixed $num An integer or its string representation
     *                   that need to be split
     *
     * @return array  Groups of three-digit numbers.
     * @access private
     * @author Kouber Saparev <kouber@php.net>
     * @since  PHP 4.2.3
     */
    private function splitNumber($num)
    {
        if (is_string($num)) {
            $ret    = array();
            $strlen = strlen($num);
            $first  = substr($num, 0, $strlen%3);

            preg_match_all('/\d{3}/', substr($num, $strlen%3, $strlen), $m);
            $ret =& $m[0];

            if ($first) {
                array_unshift($ret, $first);
            }

            return $ret;
        }
        return explode(' ', number_format($num, 0, '', ' ')); // a faster version for integers
    }

    /**
     * Converts a three-digit number to its word representation
     * in French language.
     *
     * @param integer $num  An integer between 1 and 999 inclusive.
     * @param boolean $last A flag, that determines if it is the last group of digits -
     *                      this is used to accord the plural suffix of the "hundreds".
     *                      Example: 200 = "deux cents", but 200000 = "deux cent mille".
     *
     * @return string   The words for the given number.
     * @access private
     * @author Kouber Saparev <kouber@php.net>
     */
    private function showDigitsGroup($num, $last = false)
    {
        $ret = '';

        // extract the value of each digit from the three-digit number
        $e = $num%10;                  // ones
        $d = ($num-$e)%100/10;         // tens
        $s = ($num-$d*10-$e)%1000/100; // hundreds

        // process the "hundreds" digit.
        if ($s) {
            if ($s>1) {
                $ret .= $this->_digits[$s].$this->_sep.$this->_misc_numbers[100];
                if ($last && !$e && !$d) {
                    $ret .= $this->_plural;
                }
            } else {
                $ret .= $this->_misc_numbers[100];
            }
            $ret .= $this->_sep;
        }

        // process the "tens" digit, and optionally the "ones" digit.
        if ($d) {
            // in the case of 1, the "ones" digit also must be processed
            if ($d==1) {
                if ($e<=6) {
                    $ret .= $this->_misc_numbers[10+$e];
                } else {
                    $ret .= $this->_misc_numbers[10].'-'.$this->_digits[$e];
                }
                $e = 0;
            } elseif ($d>5) {
                if ($d<8) {
                    $ret .= $this->_misc_numbers[60];

                    $resto = $d*10+$e-60;
                    if ($e==1) {
                        $ret .= $this->_sep.$this->_and.$this->_sep;
                    } elseif ($resto) {
                        $ret .= $this->_dash;
                    }

                    if ($resto) {
                        $ret .= $this->showDigitsGroup($resto);
                    }
                    $e = 0;
                } else {
                    $ret .= $this->_digits[4].$this->_dash.$this->_misc_numbers[20];

                    $resto = $d*10+$e-80;
                    if ($resto) {
                        $ret .= $this->_dash;
                        $ret .= $this->showDigitsGroup($resto);

                        $e = 0;
                    } else {
                        $ret .= $this->_plural;
                    }
                }
            } else {
                $ret .= $this->_misc_numbers[$d*10];
            }
        }

        // process the "ones" digit
        if ($e) {
            if ($d) {
                if ($e==1) {
                    $ret .= $this->_sep.$this->_and.$this->_sep;
                } else {
                    $ret .= $this->_dash;
                }
            }
            $ret .= $this->_digits[$e];
        }

        // strip excessive separators
        $ret = rtrim($ret, $this->_sep);

        return $ret;
    }

    /**
     * Converts a number to its word representation
     * in French language.
     *
     * @param integer $num An integer (or its string representation) between 9.99*-10^302
     *                        and 9.99*10^302 (999 centillions) that need to be converted to words
     *
     * @return string  The corresponding word representation
     * @access protected
     * @author Kouber Saparev <kouber@php.net>
     * @since  Numbers_Words 0.16.3
     */
    function getWords($num = 0, $power = null, $powsuffix = null)
    {
        $ret = '';

        // check if $num is a valid non-zero number
        if (!$num || preg_match('/^-?0+$/', $num) || !preg_match('/^-?\d+$/', $num)) {
            return $this->_zero;
        }

        // add a minus sign
        if (substr($num, 0, 1) == '-') {
            $ret = $this->_minus . $this->_sep;
            $num = substr($num, 1);
        }

        // if the absolute value is greater than 9.99*10^302, return infinity
        if (strlen($num)>306) {
            return $ret . $this->_infinity;
        }

        // strip excessive zero signs
        $num = ltrim($num, '0');

        // split $num to groups of three-digit numbers
        $num_groups = $this->splitNumber($num);

        $sizeof_numgroups = count($num_groups);

        foreach ($num_groups as $i=>$number) {
            // what is the corresponding exponent for the current group
            $pow = $sizeof_numgroups-$i;

            // skip processment for empty groups
            if ($number!='000') {
                if ($number!=1 || $pow!=2) {
                    $ret .= $this->showDigitsGroup($number, $i+1==$sizeof_numgroups||$pow>2).$this->_sep;
                }
                $ret .= $this->_exponent[($pow-1)*3];
                if ($pow>2 && $number>1) {
                    $ret .= $this->_plural;
                }
                $ret .= $this->_sep;
            }
        }

        return rtrim($ret, $this->_sep);
    }

    public function toCurrencyWords($int_curr, $decimal, $fraction = false, $convert_fraction = true)
    {
    }
}
